README.md
```
              _
  ___ ___ _ _|_|___ ___
 |  _| .'|_'_| |_ -|_ -|
 |_| |__,|_,_|_|___|___|
  iss@raxiss(c)2020-2023

```

# MOS6502 compiler benchmark (WIP).

---

This is work-in-progress and frequenty updates are possible.
All feedback, bug reporting and opinions are highly appreciated!
To discuss use pull-requests, issue tracking or [Defence-force](https://forum.defence-force.org/viewtopic.php?t=2501) forum.

Click to see: [Latest Benchmark Results](https://raxiss.com/compiler-benchmark/bench.html).

---

### Goal:
  Simple and fair benchmark for C cross compilers targeting MOS6502 processor.

### C Compilers:
  * ` cc65             `
  * ` gcc-6502-bits    `
  * ` kickc            `
  * ` lcc65            `
  * ` llvm-mos         `
  * ` osdk             `
  * ` sdcc-m6502       `
  * ` vbcc             `

---
#### Requirements
  * Host: Linux
  * Tools: gcc, make, cmake, lua, bash/sh
  * Compilers: open source only, so they can be build for host OS

#### Restrictions for included samples:
  * Standard C code
  * Free open source

#### Installation

###### Build and install included 6502 emulator mos6502vm:
```
  $ mkdir build
  $ cd build
  $ cmake ..
  $ make all install
```

###### Copy or symlink compiler toolschains in `bin/` directory:
```
  bin/
  ├── cc65
  │   ├── bin
  │   │   ├── cc65 ...
  ...
  ├── gcc-6502
  │   ├── bin
  │   │   ├── 6502-gcc ...
  ...
  ├── kickc
  │   ├── bin
  │   │   └── kickc.sh
  ...
  ├── llvm-mos
  │   ├── bin
  │   │   ├── clang ...
  │   │   ├── llvm-as ...
  │   │   ├── mos-clang -> clang ...
  ...
  ├── osdk-lcc65
  │   ├── bin
  │   │   ├── compiler
  │   │   ├── cpp ...
  ...
  ├── sdcc
  │   ├── bin
  │   │   ├── sdcc ...
  ...
  ├── vbcc
  │   ├── bin
  │   │   ├── vasm6502_oldstyle
  │   │   ├── vbcc6502
  │   │   ├── vc
  │   │   ├── vcpr6502
  │   │   ├── vlink ...
  ...
  └── mos6502vm (built as described above)
```

##### Running benchmarks
```
  $ cd playground
  $ ./run-all-benchmarks.sh
```
  Results are stored in `playground/www/` directory (JS files).
  Load `bench.html` in browser for visual representation.
  See [HACKING.md](playground/HACKING.md) for details about benchmarks.

##### Credits
  * Mike Chambers for [Fake6502 CPU emulator core v1.1](https://github.com/omarandlorraine/fake6502)


ChangeLog.txt
====================================
MOS6502 compiler benchmark ChangeLog
====================================

Online results at: https://raxiss.com/compiler-benchmark/bench.html

(2023-06-13)
------------
  * Use -Os for speed and -Oz for size on llvm-mos. (issue: #4 , commit: 10bb15abc58fc992871990003dd13195862676a0)
  * Added -flto to llvm-mos. (issue #5)

(2023-02-19)
------------
  * Improved mos6502vm run-time disassemby to stderr
  * Added new examplr (eight-queens)

(2023-02-17)
------------
  * Added sample 0xcafe to test binary litterals (i.e. 0b00000000).
  * Show the size of basic types
  * Samples reordered

(2023-02-15)
------------
  * radix-sort sample fixed.

(2023-02-14)
------------
  * 2 sets of data bytes+cycles each depend on optimization flags
  * Added display of optimization flags used during compilation (size/speed)
  * Added download link to compilation intermediate files: 'bench-data-YYYY-MM-DD.tar.bz2'
  * Larger dataset for sort algos - 256 bytes
  * Added new test example 'shuffle'. It's output is used as input for all sort examples
  * Added 'Legend' and some explanations of the results.

(2023-02-12)
------------
  * Initial release


playground/www/bench.js
var compilers = [
  'cc65',
  'gcc-6502',
  'kickc',
  'llvm-mos',
  'osdk-lcc65',
  'sdcc',
  'vbcc',
];

var benches = [
  'type-sizes',
  'dummy',
  'hello-world',
  'bytecpy',
  'memcopy',
  '0xcafe',
  'sieve',
  'aes256',
  'mandelbrot',
  'frogmove',
  'pi',
  'shuffle',
  'bubble-sort',
  'selection-sort',
  'insertion-sort',
  'merge-sort',
  'quick-sort',
  'counting-sort',
  'radix-sort',
  'shell-sort',
  'heap-sort',
  'eight-queens',
];

var resultsOpt = 'size';
var resultsMode = 'size';

var resultsOptHTML = {
  'size': '<a class="bo">Optimized for: size</a>',
  'speed': '<a class="bo">Optimized for: speed</a>',
};

var resultsModeHTML = {
  'size': '<a class="bo">size,</a> <a class="it">bytes</a>',
  'speed': '<a class="bo">time,</a> <a class="it">cpu cycles</a>',
};

var resultsModeRelHTML = {
  'size': '<a class="bo">size,</a>',
  'speed': '<a class="bo">time,</a>',
};

var resultsRelative = false;
var resultsRelativeBase = 0;
var dataNumbers = [];

function updateResults(mode) {
  resultsMode = mode;
  resultsRelative = false;

  var data;
  var opt;
  if(resultsOpt === "size") {
      data = bench_size;
      opt = opt_size;
  }
  if(resultsOpt === "speed") {
    data = bench_speed;
    opt = opt_speed;
  }

  document.getElementById('th_10').innerHTML = resultsModeHTML[resultsMode];

  dataNumbers = [];
  for(var i=0;i<benches.length;i++) {
    var idx = '00'+i;
    idx = 'tr_'+idx.substr(idx.length-2)
    var tr = document.getElementById(idx);
    var dat = data[benches[i]][resultsMode];
    tr.cells[0].innerText = benches[i];

    for(var j=1; j<tr.cells.length; j++) {
      var num = dat[j-1];
      var cell = tr.cells[j];
      dataNumbers.push(num);
      if('type-sizes' === benches[i]) {
        cell.innerHTML = num;
        cell.style['font-size'] = '14px';
        cell.style['line-height'] = '14px';
      } else {
        cell.innerText = num;
        if(!(Number(num)===num)) cell.style['color'] = 'darkred';
        else cell.style['color'] = '';
      }
    }
  }
}

function toggleResults() {
  updateResults((resultsMode === "size")?'speed':'size');
}

function updateResultsRel(comp) {
  if(resultsRelative && resultsRelativeBase === comp) updateResults(resultsMode);
  else {
    resultsRelative = true;
    resultsRelativeBase = comp;

    document.getElementById('th_10').innerHTML = resultsModeRelHTML[resultsMode]
      +'&nbsp;<a class="it""> (vs. '+compilers[comp]+')</a>';

    for(var i=0;i<benches.length;i++) {
      var idx = '00'+i;
      idx = 'tr_'+idx.substr(idx.length-2)
      var tr = document.getElementById(idx);

      var ref = dataNumbers[i*compilers.length+comp];
      for(var j=1; j<tr.cells.length; j++) {
        var cell = tr.cells[j];
        if('type-sizes' !== benches[i]) {
          var num = dataNumbers[i*compilers.length+j-1]/ref;
          cell.innerText = num.toFixed(comp===j-1? 0:2);
          if(!((Number(num)===num)&&(Number(ref)===ref))) cell.style['color'] = '#ebebeb';
          else cell.style['color'] = (num<1)?'darkgreen':(num==1)? '':'darkred';
        }
      }
    }
  }
}

function updateOpt(mode) {
  resultsOpt = mode;
  var date = "size" === resultsOpt? date_size : date_speed;
  var title = document.getElementById('title');
  title.innerHTML = title.innerHTML.replaceAll(/@/g,date)
  document.getElementById('th_00').innerHTML = resultsOptHTML[resultsOpt];

  var opt = "size" === resultsOpt? opt_size : opt_speed;
  for(var i=0;i<compilers.length;i++)
    document.getElementById('op_'+i).innerText = opt[compilers[i]];

  updateResults(resultsMode);
}

function toggleOpt() {
  updateOpt((resultsOpt === "size")?'speed':'size');
}

function setupPage() {
  var date = "size" === resultsOpt? date_size : date_speed;
  var arc = document.getElementById('archive').innerHTML;
  document.getElementById('archive').innerHTML = arc.replaceAll(/@/g,date);
  updateOpt('size');
}
